server/portal/libs/googledrive/operations.py

```diff
@@ -0,0 +1,179 @@
+    reached_end = not bool(scroll_token)
+    folder_mimetype = 'application/vnd.google-apps.folder'
+    listing = list(map(lambda f: {
+        'system': 'googledrive',
+        'type': 'dir' if f['mimeType'] == folder_mimetype else 'file',
+        'format': 'folder' if f['mimeType'] == folder_mimetype else 'raw',
+        'mimeType': f['mimeType'],
+        'path': f['id'],
+        'name': f['name'],
+        'length': int(f['size']) if 'size' in f.keys() else 0,
+        'lastModified': f['modifiedTime'],
+        '_links': {
+            'self': {'href': f['webViewLink']}
+        }
+    }, listing))
+
+    return {'listing': listing,
+            'nextPageToken': scroll_token,
+            'reachedEnd': reached_end}
+
+
+def search(client, system, path, offset=None, limit=100, nextPageToken=None,
+           query_string='', *args, **kwargs):
+    if not path:
+        path = 'root'
+    fields = ("mimeType, name, id, modifiedTime, "
+              "fileExtension, size, parents, webViewLink")
+    listing_call = client.files()\
+        .list(q=("'{path}' in parents and "
+                 "trashed=False and "
+                 "name contains '{query_string}'")
+              .format(path=path, query_string=query_string),
+              fields="files({}), nextPageToken"
+              .format(fields), pageSize=limit, pageToken=nextPageToken)\
+        .execute()
+    listing = listing_call.get('files')
+    scroll_token = listing_call.get('nextPageToken')
+    reached_end = not bool(scroll_token)
+    folder_mimetype = 'application/vnd.google-apps.folder'
+    listing = list(map(lambda f: {
+        'system': 'googledrive',
+        'type': 'dir' if f['mimeType'] == folder_mimetype else 'file',
+        'format': 'folder' if f['mimeType'] == folder_mimetype else 'raw',
+        'mimeType': f['mimeType'],
+        'path': f['id'],
+        'name': f['name'],
+        'length': int(f['size']) if 'size' in f.keys() else 0,
+        'lastModified': f['modifiedTime'],
+        '_links': {
+            'self': {'href': f['webViewLink']}
+        }
+    }, listing))
+
+    return {'listing': listing,
+            'nextPageToken': scroll_token,
+            'reachedEnd': reached_end}
+
+
+def iterate_listing(client, system, path, limit=100):
+    if not path:
+        path = 'root'
+    scroll_token = None
+    while True:
+        _listing = listing(client, system, path, limit=limit,
+                           nextPageToken=scroll_token)
+
+        scroll_token = _listing['nextPageToken']
+        yield from _listing['listing']
+
+        if not scroll_token:
+            break
+
+
+def walk_all(client, system, path, limit=100):
+    if not path:
+        path = 'root'
+    for f in iterate_listing(client, system, path, limit):
+        yield f
+        if f['format'] == 'folder':
+            yield from walk_all(client, system, f['path'], limit)
+
+
+def upload(client, system, path, uploaded_file, *args, **kwargs):
```

Comment:
I'm not able to select the "Google Drive" system to transfer files in this PR, is that a known issue being fixed elsewhere?
<img width="291" alt="Screen Shot 2020-12-16 at 3 50 00 PM" src="https://user-images.githubusercontent.com/20326896/102413177-27417400-3fba-11eb-811f-4e40883a4c39.png">

Reply:
fixed with the correct setting in `_PORTAL_DATAFILES_STORAGE_SYSTEMS`
```
    {
        'name': 'Google Drive',
        'system': 'googledrive',
        'scheme': 'private',
        'api': 'googledrive',
        'icon': None
    }
    ```